File: proj3/Project3-Answer-Tests/task1.py
#!/usr/bin/python
import binascii
import sys
import time
from des import *

def bit2str(s):
    t = ''.join(chr(int(s[i:i+8], 2)) for i in xrange(0, len(s), 8))
    return t

def hex2bin(s):
    return s.decode('hex')

def str2bit(s):
    bitArray = []
    for i in s:
        temp = str(bin(ord(i)))[2:]
        if len(temp) != 8:
            bitArray.extend([0] * (8 - len(temp)))
        for j in temp:
            bitArray.append(int(j))

    return bitArray

def test():
    key1 = b"\0\0\0\0\0\0\0\0"
    key2 = b"\0\0\0\0\0\0\0\2"
    message1 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
    message2 = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1]
    test_des(key1, message1)
    test_des(key1, message2)
    test_des(key2, message1)
    test_des(key2, message2)

def test_des(key, message):
    k = des(key)
    c = k.des_encrypt(message)

    #print bintohex("".join([str(e) for e in c]))

def cbc_encrypt(message, key, iv):
    """
    Args:
      message: string, bytes, cannot be unicode
      key: string, bytes, cannot be unicode
    Returns:
      ciphertext: string
    """
    # TODO: Add your code here.
    #test()


    # do padding
    numBytes = len(message)

    if numBytes % 8 != 0:
        # message[-1] += 1
        # message[-1] << (8 - (len(bin(ord(message[-1]))) - 2))
	print numBytes
	message += chr(2**7)
        message += "\0" * (7 - numBytes % 8)
    else:
        message += chr(2**7) + "\0" * 7
    

    result = ""

    key = hex2bin(key.strip())
    iv = hex2bin(iv.strip())
    k = des(key)

    for i in range(0, len(message), 8):
        plaintext = message[i:i+8]

        # IV XOR Plaintext
        temp = ""
        for j in range(0, 8):
            temp += chr(ord(plaintext[j]) ^ ord(iv[j]))

        c = k.des_encrypt(str2bit(temp))
	# The following two lines are for debugging.
	s = "".join([str(e) for e in c])
	print "".join([hex(int(s[i:i+8], 2)) for i in xrange(0, len(s), 8)])

        ciphertext = bit2str("".join([str(e) for e in c]))
	print "".join([str(e) for e in c])
        iv = ciphertext
        result += ciphertext
    
    return result


def cbc_decrypt(message, key, iv):
    """
    Args:
      message: string, bytes, cannot be unicode
      key: string, bytes, cannot be unicode
    Returns:
      plaintext: string
    """

    result = ""
    key = hex2bin(key.strip())
    iv = hex2bin(iv.strip())
    k = des(key)

    for i in range(0, len(message), 8):
        ciphertext = message[i:i+8]

        p = k.des_decrypt(str2bit(ciphertext))
        temp = bit2str("".join([str(e) for e in p]))

        # IV XOR Plaintext
        plaintext = ""
        for j in range(0, 8):
            plaintext += chr(ord(temp[j]) ^ ord(iv[j]))

        iv = ciphertext

        result += plaintext

    bits = str2bit(result)
    for i in range(len(bits) - 1, -1, -1):
        if bits[i] == 1:
            break
    return result[:i/8]


def main(argv):
    if len(argv) != 5:
        print 'Wrong number of arguments!\npython task1.py $MODE $INFILE $KEYFILE $IVFILE $OUTFILE'
        sys.exit(1)
    mode = argv[0]
    infile = argv[1]
    keyfile = argv[2]
    ivfile = argv[3]
    outfile = argv[4]
    message = None
    key = None
    iv = None
    try:
        message = open(infile, 'r').read()
        key = open(keyfile, 'r').read()
        iv = open(ivfile, 'r').read()
    except:
        print 'File Not Found'
    start = time.time()
    if mode == "enc":
        output = cbc_encrypt(message, key, iv)
    elif mode == "dec":
        output = cbc_decrypt(message, key, iv)
    else:
        print "Wrong mode!"
        sys.exit(1)
    end = time.time()
    print "Consumed time=", (end - start)
    open(outfile, 'w').write(output)

if __name__=="__main__":
    main(sys.argv[1:])

File: proj3/Project3-Answer-Tests/task1.java
import java.io.IOException;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.lang.management.ManagementFactory;
import java.lang.management.ThreadMXBean;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.charset.Charset;
import java.security.InvalidKeyException;
import java.util.Arrays;
import javax.xml.bind.DatatypeConverter;


public class task1 {
	public static byte[] padding = new byte[]{(byte)128, 0, 0, 0, 0, 0, 0, 0};

	public static String toHexString(byte[] array) {
		return DatatypeConverter.printHexBinary(array);
	}

	public static byte[] toByteArray(String s) {
		return DatatypeConverter.parseHexBinary(s);
	}
	
	public static byte[] arraySub(byte[]a, int start, int end) {
		byte[] result = new byte[end - start];
		System.arraycopy(a, start, result, 0, end - start);
		return result;
	}

	public static byte[] arrayConcat(byte[] a, byte[] b) {
		if (a == null) {
			if (b == null) {
				return null;
			} else {
				return b;
			}
		} else if (b == null) {
			return a;
		}
		byte[] result = new byte[a.length + b.length];
		System.arraycopy(a, 0, result, 0, a.length);
		System.arraycopy(b, 0, result, a.length, b.length);
		return result;
	}

	public static byte[] cbc_encrypt(byte[] message, String key, String iv) {
		// TODO: Add your code here.
		// Padding
		int messageSize = message.length;
		System.out.println(toHexString(message));
		if (messageSize % 8 != 0) {
			System.out.println(toHexString(Arrays.copyOfRange(padding, 0, 8 - (messageSize % 8))));
			message = arrayConcat(message, Arrays.copyOfRange(padding, 0, 8 - (messageSize % 8)));
		} else {
			message = arrayConcat(message, padding);
		}
		messageSize = message.length;

		// Initialize DES object.
		iv = iv.replace("\n", "");
		key = key.replace("\n", "");
		byte[] ivBytes = toByteArray(iv);
		byte[] keyBytes = toByteArray(key);

		DES des = new DES();
		Object k;
		byte[] result = null;
		try {
			k = des.makeKey(keyBytes, des.KEY_SIZE);
			for (int i=0; i<messageSize; i+=8) {
				// IV Xor Plaintext
				System.out.println(i);
				System.out.println(i+8);
				// System.out.println(message.substring(i, i+8));
				byte[] temp = Arrays.copyOfRange(message, i, i+8);
				System.out.println("before");
				System.out.println(toHexString(temp));

				for (int j=0; j<8; j++) {
					temp[j] = (byte) ( ivBytes[j] ^ temp[j] );
				}
				byte[] output = toByteArray(des.encrypt(k, temp));
				System.out.println("after");
				System.out.println(toHexString(output));

				ivBytes = output;
				System.out.println(ivBytes);
				result = arrayConcat(result, output);
			}
		} catch (InvalidKeyException e) {
			System.out.println("Invalid Key.");
		}
		return result;
	}

	public static byte[] cbc_decrypt(byte[] message, String key, String iv) {
		// TODO: Add your code here.
		int messageSize = message.length;
		// Initialize DES object.
		iv = iv.replace("\n", "");
		key = key.replace("\n", "");
		byte[] ivBytes = toByteArray(iv);
		byte[] keyBytes = toByteArray(key);

		DES des = new DES();
		Object k;
		byte[] result = null;
		try {
			k = des.makeKey(keyBytes, des.KEY_SIZE);
			for (int i=0; i<messageSize; i+=8) {
				// IV Xor Plaintext
				System.out.println(i);
				System.out.println(i+8);
				// System.out.println(message.substring(i, i+8)); 
				byte[] temp = Arrays.copyOfRange(message, i, i+8);
				System.out.println("before");
				System.out.println(toHexString(temp));

				byte[] output = toByteArray(des.decrypt(k, temp));
				for (int j=0; j<8; j++) {
					output[j] = (byte) ( ivBytes[j] ^ output[j] );
				}				
				System.out.println("after");
				System.out.println(toHexString(output));
				ivBytes = temp;
				System.out.println(ivBytes);
				result = arrayConcat(result, output);
			}
		} catch (InvalidKeyException e) {
			System.out.println("Invalid Key.");
		}
		// Discard padding.
		int result_index = result.length-1;
		byte lastByte = result[result_index];
		while (lastByte != padding[0]) {
			result_index --;
			lastByte = result[result_index];			
		}
		result = arraySub(result, 0, result_index);
		return result;
	}

	public static void main (String[] args) {
		if (args.length != 5) {
			System.out.println("Wrong number of arguments!\njava task1 $MODE $INFILE $KEYFILE $IVFILE $OUTFILE.");
			System.exit(1);
		} else {
			String mode = args[0];
			String infile = args[1];
			String keyfile = args[2];
			String ivfile = args[3];
			String outfile = args[4];
			byte[] input = readFromFile(infile);
			String key = readFromFile(keyfile, Charset.defaultCharset());
			String iv = readFromFile(ivfile, Charset.defaultCharset());
			byte[] output = null;

			double start = getCpuTime();
			// Calculate the CPU cycles.
			if (mode.equals("enc")) {
				output = cbc_encrypt(input, key, iv);
			} else if (mode.equals("dec")) {
				output = cbc_decrypt(input, key, iv);
			} else {
				System.out.println(mode);
				System.out.println("Wrong mode!");
				System.exit(1);
			}
			double end = getCpuTime();
			System.out.printf("Consumed CPU time=%f\n", end - start);
			writeToFile(outfile, output);
		}
	}

	static String readFromFile(String path, Charset encoding) {
		try {
			byte[] encoded = Files.readAllBytes(Paths.get(path));
			return new String(encoded, encoding);
		} catch (IOException e) {
			System.out.println("File Not Found.");
			return null;
		}
	}

	static byte[] readFromFile(String path) {
		try {
			byte[] encoded = Files.readAllBytes(Paths.get(path));
			return encoded;
		} catch (IOException e) {
			System.out.println("File Not Found.");
			return null;
		}
	}

	static void writeToFile(String path, byte[] data) {
		try {
			Files.write(Paths.get(path), data);
		} catch (FileNotFoundException e) {
			System.out.println("File Not Found.");
		} catch (IOException e) {
			System.out.println("File Not Found.");
		}
	}

	// Helper functions.
	private static double getCpuTime () {
		ThreadMXBean bean = ManagementFactory.getThreadMXBean();
		// getCurrentThreadCpuTime() returns the total CPU time for the current thread in nanoseconds.
		return bean.isCurrentThreadCpuTimeSupported() ? ((double)bean.getCurrentThreadCpuTime() / 1000000000): 0L;
	}

	static void testDES(String key, String message) {
		DES des = new DES();
		Object k;
		try {
			k = des.makeKey(key.getBytes(), des.KEY_SIZE);
			String output = des.encrypt(k, message.getBytes());
			// suppress output
			// System.out.println(output);
		} catch (InvalidKeyException e) {
			System.out.println("Invalid Key.");
		}
	}
	
	static void test() {
		// This function is for test and illustration purpose.
		char[] chars1 = new char[8];
		char[] chars2 = new char[8];
		Arrays.fill(chars1, '\0');
		Arrays.fill(chars2, '\0');
		String key = new String(chars1);
		String message = new String(chars2);
		testDES(key, message);
		chars2[7] = '\1';
		message = new String(chars2);
		testDES(key, message);
		chars1[7] = '\2';
		chars2[7] = '\0';
		key = new String(chars1);
		message = new String(chars2);
		testDES(key, message);
		chars2[7] = '\1';
		message = new String(chars2);
		testDES(key, message);		
	}
}
